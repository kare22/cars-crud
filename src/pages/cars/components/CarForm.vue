<template>
    <ValidationObserver ref="observer" v-slot="{ handleSubmit }">
        <BForm class="text-left text-info" @submit.stop.prevent="save">
        <div>
            <div class="my-3" v-if="isNew">
                <div class="ml-2">
                    Car category:
                </div>
                <div>
                    <ValidationProvider name="Category" :rules="{ required: true}" v-slot="validationContext">
                        <BFormSelect
                            v-model="carData.category"
                            value-field="id"
                            text-field="title"
                            :options="categories"
                            :state="getValidationState(validationContext)"
                        >
                            <template #first>
                                <b-form-select-option :value="null" disabled>-- Please select a category --</b-form-select-option>
                            </template>
                        </BFormSelect>
                        <BFormInvalidFeedback>{{ validationContext.errors[0] }}</BFormInvalidFeedback>
                    </ValidationProvider>
                </div>
            </div>
            <div class="ml-2 d-flex justify-content-between">
                <div class="mt-auto">
                    Car title:
                </div>
                <div>
                    <a
                        v-if="!isNew"
                        class="btn btn-link text-info"
                        @click="remove"
                    >
                        <BIconX class="icon" />
                    </a>
                </div>
            </div>
            <ValidationProvider name="Title" :rules="{ required: true, min: 4 }" v-slot="validationContext">
                <BFormInput
                    v-model="carData.title"
                    :state="getValidationState(validationContext)"
                ></BFormInput>
                <BFormInvalidFeedback>{{ validationContext.errors[0] }}</BFormInvalidFeedback>
            </ValidationProvider>
        </div>
        <div class="my-4">
            <div class="ml-2">
                Car description:
            </div>
            <ValidationProvider name="Description" :rules="{ required: true}" v-slot="validationContext">
                <BFormTextarea
                    v-model="carData.description"
                    :state="getValidationState(validationContext)"
                ></BFormTextarea>
                <BFormInvalidFeedback>{{ validationContext.errors[0] }}</BFormInvalidFeedback>
            </ValidationProvider>
        </div>
        <div class="row my-md-4">
            <div class="col-6 col-md-4">
                Probability
                <BFormSelect
                    v-model="carData.probability"
                    value-field="id"
                    text-field="title"
                    :options="probabilities"
                />
            </div>
            <div class="col-6 col-md-4">
                Impact
                <BFormSelect
                    v-model="carData.impact"
                    value-field="id"
                    text-field="title"
                    :options="impacts"
                />
            </div>
            <div class="col-12 col-md-4 my-3 my-md-0">
                TODO
            </div>
        </div>
        <div class="mb-4">
            <div class="ml-2">
                Car response:
            </div>
            <textarea
                v-model="carData.response"
                type="text"
                class="form-control"
            />
        </div>
        <div class="mb-3 text-right">
            <button
                v-if="isNew"
                class="btn btn-outline-info mr-3"
                @click="cancel"
            >
                Cancel
            </button>
            <button
                class="btn btn-info"
                type="submit"
            >
                Save
            </button>
        </div>
    </BForm>
    </ValidationObserver>
</template>

<script>
import { mapGetters } from 'vuex';

export default {
    name: 'CarForm',
    props: {
        car: {
            type: Object,
            default: null,
        },
    },
    data() {
        return {
            carData: {
                id: null,
                title: '',
                description: '',
                response: '',
                category: null,
                probability: null,
                impact: null,
            },
        };
    },
    computed: {
        ...mapGetters({
            probabilities: 'probabilities/data',
            impacts: 'impacts/data',
            categories: 'categories/data',
        }),
        isNew() {
            return !this.car?.id;
        },
    },
    created() {
        if(!this.isNew) {
            Object.keys(this.carData).forEach(key => {
                this.carData[key] = this.car[key];
            });
        }
    },
    methods: {
        getValidationState({ dirty, validated, valid = null }) {
            return dirty || validated ? valid : null;
        },
        async save() {
            if (!(await this.$refs?.observer?.validate())) {
                return;
            }
            if (this.isNew) {
                await this.$store.dispatch('cars/create', {car: this.carData});
            } else {
                await this.$store.dispatch('cars/update', {car: this.carData});
            }
            this.$emit('save');
        },
        cancel() {
            this.$emit('cancel');
        },
        remove() {
            this.$confirm({
                title: 'Are you sure?',
                message: `Are you sure you want to remove ${this.car.title}?`,
                button: {
                    yes: 'Yes',
                    no: 'Cancel'
                },
                callback: async(confirm) => {
                    if (confirm) {
                        await this.$store.dispatch('cars/remove', { id: this.car.id });
                        this.$toastr.e(`${this.car.title} was deleted!`);
                    }
                }
            });
        },
    },
};
</script>

<style scoped>

</style>