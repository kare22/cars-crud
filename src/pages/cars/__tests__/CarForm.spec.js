import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import { extend } from 'vee-validate';
import { min, required } from 'vee-validate/dist/rules';

import testData from './__mocks__/test-data';
import CarForm from '@/pages/cars/components/CarForm';

jest.mock('@/store');

describe('CarForm', () => {
    const mocks = {};

    let actions;
    let getters;
    let store;

    const localVue = createLocalVue();
    localVue.use(Vuex);
    localVue.use(VueRouter);
    const router = new VueRouter();

    const car = testData.cars[1][0];

    function getComponent(options = {}) {
        return mount(CarForm, {
            ...options,
            mocks,
            store,
            localVue,
            router,
        });
    }

    beforeEach(() => {
        extend('required', required);
        extend('min', min);

        actions = {
            'cars/get': () => [],
            'categories/get': () => [],
            'probabilities/get': () => [],
            'impacts/get': () => [],
        };
        getters = {
            'cars/categorized': () => testData.cars,
            'categories/data': () => testData.categories,
            'impacts/data': () => testData.impacts,
            'probabilities/data': () => testData.probabilities,
        };
        store = new Vuex.Store({
            actions,
            getters,
        });
    });

    it('snapshot test', () => {
        expect(getComponent()).toMatchSnapshot();
    });

    it('should call remove when x is triggered', async() => {
        const propsData = {
            car,
        };
        const methods = {
            remove: jest.fn(),
        };
        const wrapper = await getComponent({ methods, propsData });

        wrapper.find(`#remove_car_${car.id}`).trigger('click');

        await wrapper.vm.$nextTick();

        expect(methods.remove).toHaveBeenCalled();
    });

    it('should accept valid form fields', async() => {
        const wrapper = await getComponent({
            data() {
                return {
                    carData: {
                        title: 'Some title',
                        description: 'Some description',
                        category: 2,
                    }
                };
            },
        });

        wrapper.find('form').trigger('submit');

        await wrapper.vm.$nextTick();

        expect( await wrapper.vm.$refs.observer.validate()).toBeTruthy();
    });

    it('should not accept invalid form fields', async() => {
        const wrapper = await getComponent({
            data() {
                return {
                    carData: {
                        title: 'Some title',
                        description: '',
                        category: 2,
                    }
                };
            },
        });

        wrapper.find('form').trigger('submit');

        await wrapper.vm.$nextTick();

        expect( await wrapper.vm.$refs.observer.validate()).toBeFalsy();
    });

    it('should show categories select when id is not set', async() => {
        const propsData = {
            car: {
                title: 'Some title',
                description: '',
                category: 2,
            },
        };
        const wrapper = await getComponent({ propsData });

        expect(wrapper.find('.category-select').exists()).toBeTruthy();
    });

    it('should not show categories select when id is set', async() => {
        const propsData = {
            car: {
                id: 239,
                title: 'Some title',
                description: '',
                category: 2,
            },
        };
        const wrapper = await getComponent({ propsData });

        expect(wrapper.find('.category-select').exists()).toBeFalsy();
    });
});

