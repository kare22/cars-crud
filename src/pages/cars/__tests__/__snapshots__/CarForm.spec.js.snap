// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarForm snapshot test 1`] = `
<span><form class="text-left text-info"><div class="category-select my-3"><div class="ml-2">
                Car category:
            </div> <div><span><bformselect value-field="id" text-field="title" options="[object Object],[object Object],[object Object]"></bformselect> <bforminvalidfeedback></bforminvalidfeedback></span></div>
</div>
<div>
  <div class="ml-2 d-flex justify-content-between">
    <div class="mt-auto">
      Car title:
    </div>
    <div>
      <!---->
    </div>
  </div> <span><bforminput></bforminput> <bforminvalidfeedback></bforminvalidfeedback></span>
</div>
<div class="my-4">
  <div class="ml-2">
    Car description:
  </div> <span><bformtextarea></bformtextarea> <bforminvalidfeedback></bforminvalidfeedback></span>
</div>
<div class="row my-md-4">
  <div class="col-6 col-md-4">
    Probability
    <bformselect value-field="id" text-field="title" options="[object Object],[object Object],[object Object]"></bformselect>
  </div>
  <div class="col-6 col-md-4">
    Impact
    <bformselect value-field="id" text-field="title" options="[object Object],[object Object],[object Object]"></bformselect>
  </div>
  <div class="col-12 col-md-4 my-3 my-md-0">
    TODO
  </div>
</div>
<div class="mb-4">
  <div class="ml-2">
    Car response:
  </div> <textarea type="text" class="form-control"></textarea>
</div>
<div class="mb-3 text-right"><button class="btn btn-outline-info mr-3">
    Cancel
  </button> <button type="submit" class="btn btn-info">
    Save
  </button></div>
</form>
</span>
`;
